KAFKA-12849: KIP-744 TaskMetadata ThreadMetadata StreamsMetadata as API

jlprat opened the pull request:
Implementation of [KIP-744](https://cwiki.apache.org/confluence/x/XIrOCg).

Creates new Interfaces for TaskMetadata, ThreadMetadata, and
StreamsMetadata, providing internal implementations for each of them.

Deprecates current TaskMetadata, ThreadMetadata under o.a.k.s.processor,
and SreamsMetadata under a.o.k.s.state.

Updates references on internal classes from deprecated classes to new interfaces.

Deprecates methods on KStreams returning deprecated ThreadMeatada and
StreamsMetadta, and provides new ones returning the new interfaces.

Update Javadocs referencing to deprecated classes and methods to point
to the right ones.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

jlprat committed:
KAFKA-12849: KIP-744 TaskMetadata ThreadMetadata StreamsMetadata as API

Implementation of KIP-744.

Creates new Interfaces for TaskMetadata, ThreadMetadata, and
StreamsMetadata, providing internal implementations for each of them.

Deprecates current TaskMetadata, ThreadMetadata under o.a.k.s.processor,
and SreamsMetadata under a.o.k.s.state.

Updates references on internal classes from deprecated classes to new interfaces.

Deprecates methods on KStreams returning deprecated ThreadMeatada and
StreamsMetadta, and provides new ones returning the new interfaces.

Update Javadocs referencing to deprecated classes and methods to point
to the right ones.

jlprat commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1458,8 +1457,30 @@ public void cleanUp() {
+     *
+     * @return {@link StreamsMetadata} for each {@code KafkaStreams} instances of this application
+     */
+    public Collection<StreamsMetadata> allRunningMetadata() {
```
Not so much convinced about this name, open to suggestions if you don't like it.

ableegoldman replied:
How about `allStreamsMetadata`? I think that gets at the core difference here, which is that this returns metadata for all Streams instances.

cadonna replied:
What about `metadataForAllStreamsClients()`? I think this makes it more explicit.

jlprat replied:
I decided to for for the pattern xxxForxxx to keep consistency among different changes.

It is now `metadataForAllStreamsClients` but happy to change if anyone has reasons against it.

jlprat commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1558,12 +1607,45 @@ private void processStreamThread(final Consumer<StreamThread> consumer) {
      * @return the set of {@link ThreadMetadata}.
      */
-    public Set<ThreadMetadata> localThreadsMetadata() {
+    public Set<ThreadMetadata> threadsMetadata() {
```
Similarly here, name is maybe not the best, happy to change it if there is a better alternative.

ableegoldman replied:
I think `localThreadMetadata` sounds better to me (actually even better than the original `localThreadsMetadata` tbh), but I'm fine with `threadsMetadata` too

cadonna replied:
nit: I think `localThreadMetadata` does not sound consistent since metadata of multiple stream threads is returned. What about `metadataForLocalThreads()`?

jlprat replied:
I decided to for for the pattern xxxForxxx to keep consistency among different changes.

It is now `metadataForLocalThreads` but happy to change if anyone has reasons against it.

ableegoldman replied:
@cadonna `localThreadMetadata` still sounds more correct to me than `localThreadsMetadata`. I really can't explain it other than to say that English is weird, and names/titles like this do not always follow the regular rules of grammar/plurals 🤷‍♀️ 

But actually I like your suggestion `metadataForLocalThreads()` even better than any of them, SGTM

jlprat commented on streams/src/main/java/org/apache/kafka/streams/processor/TaskMetadata.java:
```diff
@@ -27,10 +27,12 @@
 public class TaskMetadata {
 
-    private final TaskId taskId;
+    private final String taskId;
```
This reverts TaskMetadata to the state before KIP-740. As this TaskMetadata is deprecated, no need to do any type change here.

jlprat commented on streams/src/main/java/org/apache/kafka/streams/processor/TaskMetadata.java:
```diff
@@ -68,17 +61,8 @@ public TaskMetadata(final TaskId taskId,
     /**
      * @return the basic task metadata such as subtopology and partition id
      */
-    public TaskId getTaskId() {
```
Same here, this reverts to the state prior KIP-740. Users are anyway discouraged to use this class.

jlprat commented on streams/src/main/java/org/apache/kafka/streams/TaskMetadata.java:
```diff
@@ -0,0 +1,73 @@
+    /**
+     * @return the basic task metadata such as subtopology and partition id
+     */
+    TaskId taskId();
```
Cleaned up API here, with proper naming convention for the new TaskMetada

jlprat reviewed (commented):
Added some notes to help navigate the PR

jlprat commented:
As @cadonna, @ableegoldman and @guozhangwang showed some interest on the KIP, I guess maybe any of you can review this PR?

jlprat commented:
Failure was:
```
[2021-06-08T14:01:59.439Z] FAILURE: Build failed with an exception.
[2021-06-08T14:01:59.439Z] 
[2021-06-08T14:01:59.439Z] * What went wrong:
[2021-06-08T14:01:59.439Z] Execution failed for task ':core:integrationTest'.
[2021-06-08T14:01:59.439Z] > Process 'Gradle Test Executor 127' finished with non-zero exit value 1
[2021-06-08T14:01:59.439Z]   This problem might be caused by incorrect test process configuration.
[2021-06-08T14:01:59.439Z]   Please refer to the test execution section in the User Manual at https://docs.gradle.org/7.0.2/userguide/java_testing.html#sec:test_execution
```

On top of the common `RaftClusterTest` ones

jlprat commented:
Following Konstantine's recommendation on the mailing list in regards of the dates for the release of Kafka 3.0.0, I'm pinging again potential reviewers for this KIP's implementation.
@ableegoldman (as creator of the original Jira Ticket), @cadonna and @guozhangwang (as voters of the KIP), I would highly appreciate your feedback in this PR.

Thanks a lot!

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1478,8 +1499,36 @@ public void cleanUp() {
+     * @return {@link StreamsMetadata} for each {@code KafkaStreams} instances with the provide {@code storeName} of
+     * this application
+     */
+    public Collection<StreamsMetadata> allMetadataForGivenStore(final String storeName) {
```
I don't feel too strongly about any of these names, but it's probably best to try and keep them (a) consistent with each other, where possible, and (b) consistent with the return type. Assuming we go with something like `allStreamsMetadata` for the new method above, maybe something like `streamsMetadataForStore` would be appropriate for this one?

cadonna replied:
I am +1 for `streamsMetadataForStore`.

jlprat replied:
I decided to for for the pattern xxxForxxx to keep consistency among different changes.

It is now `streamsMetadataForStore` but happy to change if anyone has reasons against it.

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/TaskMetadata.java:
```diff
@@ -27,10 +27,12 @@
 
 /**
  * Represents the state of a single task running within a {@link KafkaStreams} application.
+ * @deprecated since 3.0, not intended for public use, use {@link org.apache.kafka.streams.TaskMetadata} instead.
```
```suggestion
 * @deprecated since 3.0, use {@link org.apache.kafka.streams.TaskMetadata} instead.
```
I think people might be a bit confused if we say this, just point them to the new interface

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:
```diff
@@ -229,7 +230,7 @@ private void rebuildMetadata(final Map<HostInfo, Set<TopicPartition>> activePart
             allMetadata = Collections.emptyList();
-            localMetadata.set(new StreamsMetadata(thisHost,
+            localMetadata.set(new StreamsMetadataImpl(thisHost,
                                                   Collections.emptySet(),
                                                   Collections.emptySet(),
                                                   Collections.emptySet(),
```
nit: fix alignment (here and below change)

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskMetadataImpl.java:
```diff
@@ -0,0 +1,105 @@
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        final org.apache.kafka.streams.processor.internals.TaskMetadataImpl that = (org.apache.kafka.streams.processor.internals.TaskMetadataImpl) o;
```
Does this need to be fully qualified? There's only one `TaskMetadataImpl` class right?

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/state/internals/StreamsMetadataImpl.java:
```diff
@@ -0,0 +1,163 @@
+     * operations.
+     */
+    public final static StreamsMetadataImpl NOT_AVAILABLE = new StreamsMetadataImpl(HostInfo.unavailable(),
+                                                                            Collections.emptySet(),
+                                                                            Collections.emptySet(),
+                                                                            Collections.emptySet(),
+                                                                            Collections.emptySet());
+
```
nit: fix alignment

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/StreamsMetadata.java:
```diff
@@ -0,0 +1,88 @@
+     * <pre>
+     * {@code
+     * Objects.hash(hostInfo(), stateStoreNames(), topicPartitions(), standbyStateStoreNames(), standbyTopicPartitions());
+     * }
```
missing closing tag `</pre>`?

ableegoldman reviewed (commented):
Did a quick pass, looks good for the most part but I think we can take the opportunity to clean up the method names a bit. Lmk what you think of the suggestions

cadonna commented on streams/src/main/java/org/apache/kafka/streams/TaskMetadata.java:
```diff
@@ -0,0 +1,73 @@
+    /**
+     * This function will return a set of the current TopicPartitions
+     */
+    Set<TopicPartition> topicPartitions();
```
Could you please use javadoc mark-up like `@return` and `@param` for the docs? Here and for the other methods.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/TaskMetadata.java:
```diff
@@ -27,10 +27,12 @@
  * Represents the state of a single task running within a {@link KafkaStreams} application.
+ * @deprecated since 3.0, not intended for public use, use {@link org.apache.kafka.streams.TaskMetadata} instead.
  */
+@Deprecated
```
nit: If we deprecate the class, we do not need to deprecate the constructor anymore, right?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/ThreadMetadata.java:
```diff
@@ -0,0 +1,91 @@
+    /**
+     * This function will return the set of the {@link TaskMetadata} for the current active tasks
+     */
+    Set<TaskMetadata> activeTasks();
```
Could you use javadoc mark-up for the docs? Here and for the other methods.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/StreamsMetadata.java:
```diff
@@ -0,0 +1,88 @@
+    /**
+     * This method is equivalent to call {@code StreamsMetadata.hostInfo().host();}
+     */
+    String host();
```
Could you describe the return value with `@return`? Please also check the other methods if they use appropriate javadocs mark-up.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskMetadataImpl.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Override
+    public boolean equals(final Object o) {
```
Could you add unit tests for `equals()` and `hashCode()`? Since we guarantee a behavior on the interface, it makes sense to verify if the behavior of the implementation is correct.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ThreadMetadataImpl.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    @Override
+    public boolean equals(final Object o) {
```
Could you add unit tests for equals() and hashCode()? Since we guarantee a behavior on the interface, it makes sense to verify if the behavior of the implementation is correct.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/state/internals/StreamsMetadataImpl.java:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    @Override
+    public boolean equals(final Object o) {
```
Could you add unit tests for equals() and hashCode()? Since we guarantee a behavior on the interface, it makes sense to verify if the behavior of the implementation is correct.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -1972,12 +1971,6 @@ public void shouldAlwaysUpdateTasksMetadataAfterChangingState() {
         assertEquals(StreamThread.State.RUNNING.name(), metadata.threadState());
     }
 
-    private void assertThreadMetadataHasEmptyTasksWithState(final ThreadMetadata metadata, final StreamThread.State state) {
```
Thank you for the clean up!

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskMetadataImpl.java:
```diff
@@ -0,0 +1,105 @@
+                            final Map<TopicPartition, Long> endOffsets,
+                            final Optional<Long> timeCurrentIdlingStarted) {
+        this.taskId = taskId;
+        this.topicPartitions = topicPartitions;
+        this.committedOffsets = committedOffsets;
+        this.endOffsets = endOffsets;
```
Since this class is read only, it makes sense to use `Collections.unmodifiable*()` methods here to avoid modification of the contents of the member fields.
Unit tests as in `StreamsMetadataTest` would also be good.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ThreadMetadataImpl.java:
```diff
@@ -0,0 +1,147 @@
+                              final Set<TaskMetadata> standbyTasks) {
+        this.mainConsumerClientId = mainConsumerClientId;
+        this.restoreConsumerClientId = restoreConsumerClientId;
+        this.producerClientIds = producerClientIds;
```
I think, you should use `Collections.unmodifiableSet()` also here.
Also unit tests to verify the immutability as in `StreamsMetadataTest` would be great!

cadonna commented on streams/src/main/java/org/apache/kafka/streams/state/internals/StreamsMetadataImpl.java:
```diff
@@ -0,0 +1,163 @@
+     */
+    @Override
+    public Set<String> stateStoreNames() {
+        return Collections.unmodifiableSet(stateStoreNames);
```
I think it would be better to make the collections immutable in the constructor since they should also not be modified within this class (for now).

cadonna commented on docs/streams/upgrade-guide.html:
```diff
@@ -121,10 +121,23 @@ <h3><a id="streams_api_changes_300" href="#streams_api_changes_300">Streams API
+        <code>org.apache.kafka.streams.TaskMetadata</code> to be used instead. This change was introduced to better reflect the fact that <code>TaskMetadata</code> was not meant to be instantiated outside
+        of Kafka codebase.
+        Please note that the new <code>TaskMetadata</code> offers APIs that better represent the task id as an actual <code>TaskId</code> object instead of a String. Please migrate to the new
+        <code>org.apache.kafka.streams.TaskMetadata</code> which offers these better methods, for example, by using the new <code>ThreadMetadata#getActiveTasks</code> and <code>ThreadMetadata#getStandbyTasks</code>.
```
```suggestion
        <code>org.apache.kafka.streams.TaskMetadata</code> which offers these better methods, for example, by using the new <code>ThreadMetadata#activeTasks</code> and <code>ThreadMetadata#standbyTasks</code>.
```

cadonna reviewed (commented):
@jlprat Thank you for the PR!

Sorry for the late review!

Here my feedback.

jlprat committed:
Applies suggestions regarding method names, refactors, and formatting

jlprat committed:
Add tests for equals hashcode and unmodifiable collections

jlprat committed:
Add guard against null otherwise StreamThreadTest tests were failing

jlprat commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ThreadMetadataImpl.java:
```diff
@@ -57,7 +57,11 @@ public ThreadMetadataImpl(final String threadName,
         this.mainConsumerClientId = mainConsumerClientId;
         this.restoreConsumerClientId = restoreConsumerClientId;
-        this.producerClientIds = Collections.unmodifiableSet(producerClientIds);
+        if (producerClientIds != null) {
+            this.producerClientIds = Collections.unmodifiableSet(producerClientIds);
+        } else {
+            this.producerClientIds = Collections.emptySet();
+        }
```
@cadonna, I needed to add this guard as ThreadMetadataTest was failing with NPE.

cadonna replied:
I could not reproduce this issue. the set of producer IDs should never be `null`. If this is the case the test is wrong.

cadonna replied:
OK, now I see. You meant `StreamThreadTest`. There you need to setup the mock for the `TaskManager` to return at least an empty set with `expect(taskManager.producerClientIds()).andStubReturn(Collections.emptySet());`.

jlprat commented:
Needed to rebase as there were some conflicts with trunk, hence the force push.

I applied the changes in separate commits:
- One for method renames, formattings and small refactors (like the unmodifiable collections)
- One for Adding tests
- One to fix the regression caused by the unmodifiable collections in constructor.

jlprat commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskMetadataImplTest.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    @Test
+    public void shouldFollowEqualsAndHasCodeContract() {
```
@cadonna For hash code and equals contract validation, I decided to keep it under one single test, and validate the positive cases and any of the reasons why it might fall into the negative case.
Please note that objects differing only on committed offsets, end offsets, and/or time curring idling started will be considered equals.

jlprat commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ThreadMetadataImplTest.java:
```diff
@@ -0,0 +1,221 @@
+    }
+
+    @Test
+    public void shouldFollowHashCodeAndEqualsContract() {
```
@cadonna For hash code and equals contract validation, I decided to keep it under one single test, and validate the positive case and any of the reasons why it might fall into the negative case.

cadonna replied:
I prefer to have separate tests, because when something fails you immediately know the cause. But if you like to keep like that, I am fine with it.

jlprat commented on streams/src/test/java/org/apache/kafka/streams/state/StreamsMetadataTest.java:
```diff
@@ -55,6 +62,63 @@ public void shouldNotAllowModificationOfInternalStateViaGetters() {
     }
 
+    @Test
+    public void shouldFollowHashCodeAndEqualsContract() {
```
@cadonna For hash code and equals contract validation, I decided to keep it under one single test, and validate the positive case and any of the reasons why it might fall into the negative case.

cadonna replied:
see my other comment about separate tests.

jlprat reviewed (commented):
Added some notes about the hash code and equals testing approach

jlprat commented:
@ableegoldman and @cadonna thanks a lot for your reviews! I think I addressed all of your comments. Let me know what do you think about them.

jlprat commented:
I see some usual suspects on test failures, but  `org.apache.kafka.connect.integration.SourceConnectorsIntegrationTest.testSwitchingToTopicCreationEnabled`is new. Looking at the stack trace, doesn't seem relevant to the changes made in this PR.

```
java.lang.NullPointerException
	at java.util.concurrent.ConcurrentHashMap.get(ConcurrentHashMap.java:936)
	at org.reflections.Store.getAllIncluding(Store.java:82)
	at org.reflections.Store.getAll(Store.java:93)
	at org.reflections.Reflections.getSubTypesOf(Reflections.java:404)
	at org.apache.kafka.connect.runtime.isolation.DelegatingClassLoader.getPluginDesc(DelegatingClassLoader.java:352)
	at org.apache.kafka.connect.runtime.isolation.DelegatingClassLoader.scanPluginPath(DelegatingClassLoader.java:337)
	at org.apache.kafka.connect.runtime.isolation.DelegatingClassLoader.scanUrlsAndAddPlugins(DelegatingClassLoader.java:268)
	at org.apache.kafka.connect.runtime.isolation.DelegatingClassLoader.initPluginLoader(DelegatingClassLoader.java:216)
	at org.apache.kafka.connect.runtime.isolation.DelegatingClassLoader.initLoaders(DelegatingClassLoader.java:209)
	at org.apache.kafka.connect.runtime.isolation.Plugins.<init>(Plugins.java:61)
	at org.apache.kafka.connect.cli.ConnectDistributed.startConnect(ConnectDistributed.java:93)
	at org.apache.kafka.connect.util.clusters.WorkerHandle.start(WorkerHandle.java:50)
	at org.apache.kafka.connect.util.clusters.EmbeddedConnectCluster.addWorker(EmbeddedConnectCluster.java:174)
	at org.apache.kafka.connect.integration.SourceConnectorsIntegrationTest.lambda$testSwitchingToTopicCreationEnabled$1(SourceConnectorsIntegrationTest.java:197)
	at java.util.stream.Streams$RangeIntSpliterator.forEachRemaining(Streams.java:110)
	at java.util.stream.IntPipeline$Head.forEach(IntPipeline.java:559)
	at org.apache.kafka.connect.integration.SourceConnectorsIntegrationTest.testSwitchingToTopicCreationEnabled(SourceConnectorsIntegrationTest.java:197)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at sun.reflect.GeneratedMethodAccessor3.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:121)
	at sun.reflect.GeneratedMethodAccessor2.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.lang.Thread.run(Thread.java:748)
```
Is there some relation I'm missing, or should I create a new issue for this flaky test?

cadonna commented on docs/streams/upgrade-guide.html:
```diff
@@ -121,10 +121,23 @@ <h3><a id="streams_api_changes_300" href="#streams_api_changes_300">Streams API
+    <ul>
+        <li>Users of <code>KafkaStreams#allMetadata</code> are meant to migrate to the new <code>KafkaStreams#allRunningMetadata</code>.</li>
+        <li>Users of <code>KafkaStreams#allMetadataForStore(String)</code> are meant to migrate to the new <code>KafkaStreams#allMetadataForGivenStore(String)</code>.</li>
+        <li>Users of <code>KafkaStreams#localThreadsMetadata</code> are meant to migrate to the new <code>KafkaStreams#threadsMetadata</code>.</li>
```
You need to adapt this text to the new method names.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:
```diff
@@ -229,11 +230,11 @@ private void rebuildMetadata(final Map<HostInfo, Set<TopicPartition>> activePart
+                    Collections.emptySet(),
+                    Collections.emptySet(),
+                    Collections.emptySet()
             ));
```
nit: it should be 4 not 8 spaces
```suggestion
            localMetadata.set(new StreamsMetadataImpl(
                thisHost,
                Collections.emptySet(),
                Collections.emptySet(),
                Collections.emptySet(),
                Collections.emptySet()
            ));
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java:
```diff
@@ -258,11 +259,11 @@ private void rebuildMetadata(final Map<HostInfo, Set<TopicPartition>> activePart
+                        activeStoresOnHost,
+                        activePartitionsOnHost,
+                        standbyStoresOnHost,
+                        standbyPartitionsOnHost);
```
```suggestion
                final StreamsMetadata metadata = new StreamsMetadataImpl(
                    hostInfo,
                    activeStoresOnHost,
                    activePartitionsOnHost,
                    standbyStoresOnHost,
                    standbyPartitionsOnHost
                );
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/state/internals/StreamsMetadataImpl.java:
```diff
@@ -0,0 +1,163 @@
+            Collections.emptySet(),
+            Collections.emptySet(),
+            Collections.emptySet(),
+            Collections.emptySet());
```
```suggestion
    public final static StreamsMetadataImpl NOT_AVAILABLE = new StreamsMetadataImpl(
        HostInfo.unavailable(),
        Collections.emptySet(),
        Collections.emptySet(),
        Collections.emptySet(),
        Collections.emptySet()
    );
```

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskMetadataImplTest.java:
```diff
@@ -0,0 +1,141 @@
+                TOPIC_PARTITIONS,
+                COMMITTED_OFFSETS,
+                END_OFFSETS,
+                TIME_CURRENT_IDLING_STARTED);
```
nit:
```suggestion
        taskMetadata = new TaskMetadataImpl(
            TASK_ID,
            TOPIC_PARTITIONS,
            COMMITTED_OFFSETS,
            END_OFFSETS,
            TIME_CURRENT_IDLING_STARTED
        );
```

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskMetadataImplTest.java:
```diff
@@ -0,0 +1,141 @@
+    public void shouldNotAllowModificationOfInternalStateViaGetters() {
+        assertTrue(isUnmodifiable(taskMetadata.topicPartitions()));
+        assertTrue(isUnmodifiable(taskMetadata.committedOffsets()));
+        assertTrue(isUnmodifiable(taskMetadata.endOffsets()));
```
In general, for new assertions we usually use
```suggestion
        assertThat(isUnmodifiable(taskMetadata.topicPartitions(), is(true)));
        assertThat(isUnmodifiable(taskMetadata.committedOffsets(), is(true)));
        assertThat(isUnmodifiable(taskMetadata.endOffsets()), is(true));
```
This applies here and in other places of this PR.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ThreadMetadataImplTest.java:
```diff
@@ -0,0 +1,221 @@
+    public static final TopicPartition TP_1_0 = new TopicPartition("t", 1);
+    public static final TopicPartition TP_0_1 = new TopicPartition("t", 2);
+    public static final TopicPartition TP_1_1 = new TopicPartition("t", 3);
+    public static final TaskMetadata TM_0 = new TaskMetadataImpl(
+            TASK_ID_0,
+            mkSet(TP_0_0, TP_1_0),
+            mkMap(mkEntry(TP_0_0, 1L), mkEntry(TP_1_0, 2L)),
+            mkMap(mkEntry(TP_0_0, 1L), mkEntry(TP_1_0, 2L)),
+            Optional.of(3L));
```
nit:
```suggestion
    public static final TaskMetadata TM_0 = new TaskMetadataImpl(
        TASK_ID_0,
        mkSet(TP_0_0, TP_1_0),
        mkMap(mkEntry(TP_0_0, 1L), mkEntry(TP_1_0, 2L)),
        mkMap(mkEntry(TP_0_0, 1L), mkEntry(TP_1_0, 2L)),
        Optional.of(3L)
    );
```
here and below.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/ThreadMetadata.java:
```diff
@@ -0,0 +1,96 @@
+
+    /**
+     * Compares the specified object with this ThreadMetadata. Returns {@code true} if and only if the specified object is
+     * also a TaskMetadata and both {@code threadName()} are equal, {@code threadState()} are equal, {@code activeTasks()} contain the same
```
`TaskMetadata` -> `ThreadMetadata`

cadonna commented on streams/src/test/java/org/apache/kafka/streams/state/StreamsMetadataTest.java:
```diff
@@ -18,30 +18,39 @@
-            mkSet(TP_0, TP_1),
-            mkSet("store2"),
-            mkSet(TP_1)
+        streamsMetadata = new StreamsMetadataImpl(
+                HOST_INFO,
+                STATE_STORE_NAMES,
+                TOPIC_PARTITIONS,
+                STAND_BY_STORE_NAMES,
+                STANDBY_TOPIC_PARTITIONS
```
nit: The indentation was actually correct before.

cadonna reviewed (commented):
@jlprat Thank you for the update!

Here my comments!

Regarding the docs of the interfaces, could you provide at least one initial sentence for each javadoc? I know that the old classes were not documented that well. If you do not have time, we can also postpone the improvements of the docs to a separate PR and get this PR merged as soon as possible.

jlprat commented:
Hi @cadonna I'll try to fix address the comments either today or tomorrow.

Regarding the `SourceConnectorsIntegrationTest` failing in Jenkins, do you agree that is unrelated?

cadonna commented:
> Regarding the SourceConnectorsIntegrationTest failing in Jenkins, do you agree that is unrelated?

Yes, I agree!

jlprat committed:
Applies feedback

It includes
- Split equality tests by cases
- Add JavaDoc on methods
- Fix formatting
- Use proper assertThat instead of other means
- Fix names, use the proper names consistently

jlprat commented:
Hi @cadonna I applied all the feedback you provided. I gave my best at adding some Extra sentences on the JavaDocs for the interfaces. Let me know if they suffice.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/StreamsMetadata.java:
```diff
@@ -0,0 +1,100 @@
+import java.util.Set;
+
+/**
+ * Represents the state of the different a given Kafka Streams instance running within a {@link KafkaStreams} application.
```
```suggestion
 * Metadata of a Kafka Streams client.
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/StreamsMetadata.java:
```diff
@@ -0,0 +1,100 @@
+     * instance, which is typically host/port
+     *
+     * @return {@link HostInfo} corresponding to the streams instance
+     */
```
```suggestion
    /**
     * The value of {@link StreamsConfig#APPLICATION_SERVER_CONFIG} configured for the Streams
     * client.
     *
     * @return {@link HostInfo} corresponding to the Streams client
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/StreamsMetadata.java:
```diff
@@ -0,0 +1,100 @@
+     * State stores owned by the instance as an active replica
+     *
+     * @return set of active state store names
+     */
```
```suggestion
    /**
     * Names of the state stores assigned to active tasks of the Streams client.
     *
     * @return names of the state stores assigned to active tasks
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/StreamsMetadata.java:
```diff
@@ -0,0 +1,100 @@
+     * Topic partitions consumed by the instance as an active replica
+     *
+     * @return set of active topic partitions
+     */
```
```suggestion
    /**
     * Source topic partitions of the active tasks of the Streams client.
     *
     * @return source topic partitions of the active tasks
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/StreamsMetadata.java:
```diff
@@ -0,0 +1,100 @@
+     * (Source) Topic partitions for which the instance acts as standby.
+     *
+     * @return set of standby topic partitions
+     */
```
```suggestion
    /**
     * Changelog topic partitions for the state stores the standby tasks of the Streams client replicates.
     *
     * @return set of changelog topic partitions of the standby tasks
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/StreamsMetadata.java:
```diff
@@ -0,0 +1,100 @@
+     * State stores owned by the instance as a standby replica
+     *
+     * @return set of standby state store names
+     */
```
```suggestion
    /**
     * Names of the state stores assigned to standby tasks of the Streams client.
     *
     * @return names of the state stores assigned to standby tasks
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/StreamsMetadata.java:
```diff
@@ -0,0 +1,100 @@
+     * This method is equivalent to call {@code StreamsMetadata.hostInfo().host();}
+     *
+     * @return the host where the given process runs
+     */
```
```suggestion
    /**
     * Host where the Streams client runs. 
     *
     * This method is equivalent to {@code StreamsMetadata.hostInfo().host();}
     *
     * @return the host where the Streams client runs
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/StreamsMetadata.java:
```diff
@@ -0,0 +1,100 @@
+     * This method is equivalent to call {@code StreamsMetadata.hostInfo().port();}
+     *
+     * @return the port number where the given process runs
+     */
```
```suggestion
    /**
     * Port on which the Streams client listens.
     * 
     * This method is equivalent to {@code StreamsMetadata.hostInfo().port();}
     *
     * @return the port on which Streams client listens
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/TaskMetadata.java:
```diff
@@ -0,0 +1,87 @@
+     * This function will return a {@link TaskId} with basic task metadata
+     *
+     * @return the basic task metadata such as subtopology and partition id
+     */
```
```suggestion
    /**
     * Task ID of the task.
     *
     * @return task ID consisting of subtopology and partition ID
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/TaskMetadata.java:
```diff
@@ -0,0 +1,87 @@
+     * This function will return a set of the current TopicPartitions
+     *
+     * @return set of topic partitions
+     */
```
```suggestion
    /**
     * Source topic partitions of the task.
     *
     * @return source topic partitions
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/TaskMetadata.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.Set;
+
+
+/**
+ * Represents the state of a single task running within a {@link KafkaStreams} application.
+ */
```
```suggestion
/**
 * Metadata of a task.
 */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/TaskMetadata.java:
```diff
@@ -0,0 +1,87 @@
+     */
+    Set<TopicPartition> topicPartitions();
+
+    /**
+     * This function will return a map of TopicPartitions and the highest committed offset seen so far
+     *
+     * @return map with an entry for all topic partitions with the committed offset as a value
+     */
```
```suggestion
    /**
     * Offsets of the source topic partitions committed so far by the task.
     *
     * @return map from source topic partitions to committed offsets
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/TaskMetadata.java:
```diff
@@ -0,0 +1,87 @@
+     */
+    Map<TopicPartition, Long> committedOffsets();
+
+    /**
+     * This function will return a map of TopicPartitions and the highest offset seen so far in the Topic
+     *
+     * @return map with an entry for all topic partitions with the highest offset as a value
+     */
```
```suggestion
    /**
     * End offsets of the source topic partitions of the task.
     *
     * @return map source topic partition to end offsets
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/TaskMetadata.java:
```diff
@@ -0,0 +1,87 @@
+     */
+    Map<TopicPartition, Long> endOffsets();
+
+    /**
+     * This function will return the time task idling started, if the task is not currently idling it will return empty
+     *
+     * @return A filled {@code Optional} with the time where task idling started, and empty {@code Optional} otherwise
+     */
```
```suggestion
    /**
     * Time task idling started. If the task is not currently idling it will return empty.
     *
     * @return time when task idling started, empty {@code Optional} if the task is currently not idling
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/ThreadMetadata.java:
```diff
@@ -0,0 +1,110 @@
+
+import java.util.Set;
+
+/**
+ * Represents the state of a single thread running within a {@link KafkaStreams} application.
+ */
```
```suggestion
/**
 * Metadata of a stream thread.
 */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/ThreadMetadata.java:
```diff
@@ -0,0 +1,110 @@
+public interface ThreadMetadata {
+
+
+    /**
+     * This function will return the state of the Thread
+     * @return the state of the Thread
+     */
```
```suggestion
    /**
     * State of the stream thread
     *
     * @return the state
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/ThreadMetadata.java:
```diff
@@ -0,0 +1,110 @@
+     */
+    String threadState();
+
+    /**
+     * This function will return the name of the Thread
+     *
+     * @return the name of the Thread
+     */
```
```suggestion
    /**
     * Name of the stream thread
     *
     * @return the name
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/ThreadMetadata.java:
```diff
@@ -0,0 +1,110 @@
+     */
+    String threadName();
+
+    /**
+     * This function will return the set of the {@link TaskMetadata} for the current active tasks
+     *
+     * @return a set of metadata for the active tasks
+     */
```
```suggestion
    /**
     * Metadata of the active tasks assigned to the stream thread.
     *
     * @return metadata of the active tasks
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/ThreadMetadata.java:
```diff
@@ -0,0 +1,110 @@
+     */
+    Set<TaskMetadata> activeTasks();
+
+    /**
+     * This function will return the set of the {@link TaskMetadata} for the current standby tasks
+     *
+     * @return a set of metadata for the standby tasks
+     */
```
```suggestion
    /**
     * Metadata of the standby tasks assigned to the stream thread.
     *
     * @return metadata of the standby tasks

     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/ThreadMetadata.java:
```diff
@@ -0,0 +1,110 @@
+     */
+    Set<TaskMetadata> standbyTasks();
+
+    /**
+     * This function will return the Client Id for the consumer
+     *
+     * @return the consumer Client Id
+     */
```
```suggestion
    /**
     * Client ID of the Kafka consumer used by the stream thread.
     *
     * @return client ID of the Kafka consumer
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/ThreadMetadata.java:
```diff
@@ -0,0 +1,110 @@
+     */
+    String consumerClientId();
+
+    /**
+     * This function will return the Client id for the restore consumer
+     *
+     * @return the restore consumer Client Id
+     */
```
```suggestion
    /**
     * Client ID of the restore Kafka consumer used by the stream thread
     *
     * @return client ID of the restore Kafka consumer
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/ThreadMetadata.java:
```diff
@@ -0,0 +1,110 @@
+     */
+    String restoreConsumerClientId();
+
+    /**
+     * This function will return the set of Client Ids for the producers
+     *
+     * @return set of producer Client Ids
+     */
```
```suggestion
    /**
     * Client IDs of the Kafka producers used by the stream thread.
     *
     * @return client IDs of the Kafka producers
     */
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/ThreadMetadata.java:
```diff
@@ -0,0 +1,110 @@
+     */
+    Set<String> producerClientIds();
+
+    /**
+     * This function will return the Client Id for the admin client
+     *
+     * @return the admin Client Id
+     */
```
```suggestion
    /**
     * Client ID of the admin client used by the stream thread.
     *
     * @return client ID of the admin client
     */
```

cadonna reviewed (commented):
@jlprat I did a pass over the javadocs.

jlprat committed:
Apply suggestions from code review

Co-authored-by: Bruno Cadonna <cadonna@apache.org>

jlprat commented:
@cadonna accepted your suggestions. Thanks a bunch!a

cadonna reviewed (approved):
@jlprat, Thanks a lot!

LGTM!

cadonna commented:
I am just waiting for the builds to finish and then I will merge.

jlprat commented:
It seems that the test failures were RaftClusterTest and SslTransportLayerTest ones. Both groups being known flaky ones AFAIU.

cadonna commented:
The test failures are unrelated and known to be flaky:
```
Build / JDK 8 and Scala 2.12 / org.apache.kafka.common.network.SslTransportLayerTest.[1] tlsProtocol=TLSv1.2, useInlinePem=false
Build / JDK 8 and Scala 2.12 / org.apache.kafka.common.network.SslTransportLayerTest.[2] tlsProtocol=TLSv1.2, useInlinePem=true
Build / JDK 8 and Scala 2.12 / org.apache.kafka.common.network.SslTransportLayerTest.[1] tlsProtocol=TLSv1.2, useInlinePem=false
Build / JDK 8 and Scala 2.12 / org.apache.kafka.common.network.SslTransportLayerTest.[2] tlsProtocol=TLSv1.2, useInlinePem=true
Build / JDK 8 and Scala 2.12 / kafka.server.RaftClusterTest.testCreateClusterAndCreateAndManyTopics()
Build  / JDK 11 and Scala 2.13 /  org.apache.kafka.common.network.SslTransportLayerTest.[1]  tlsProtocol=TLSv1.2, useInlinePem=false
Build / JDK 11 and Scala 2.13 / org.apache.kafka.common.network.SslTransportLayerTest.[2] tlsProtocol=TLSv1.2, useInlinePem=true
Build  / JDK 11 and Scala 2.13 /  org.apache.kafka.common.network.SslTransportLayerTest.[3]  tlsProtocol=TLSv1.3, useInlinePem=false
Build  / JDK 11 and Scala 2.13 /  org.apache.kafka.common.network.SslTransportLayerTest.[1]  tlsProtocol=TLSv1.2, useInlinePem=false
Build / JDK 11 and Scala 2.13 / org.apache.kafka.common.network.SslTransportLayerTest.[2] tlsProtocol=TLSv1.2, useInlinePem=true
Build  / JDK 11 and Scala 2.13 /  org.apache.kafka.common.network.SslTransportLayerTest.[3]  tlsProtocol=TLSv1.3, useInlinePem=false
Build / JDK 11 and Scala 2.13 / kafka.server.RaftClusterTest.testCreateClusterAndCreateAndManyTopicsWithManyPartitions()
Build / JDK 11 and Scala 2.13 / kafka.server.RaftClusterTest.testCreateClusterAndCreateAndManyTopics()
```

jlprat commented:
Thanks a lot @cadonna and @ableegoldman for the reviews!
